fdbclient/NativeAPI.actor.cpp

```diff
@@ -2192,6 +2217,134 @@ ACTOR Future<Void> watchValue(Future<Version> version, Key key, Optional<Value>
+				return Void();
+			}
+		}
+	}
+}
+
+ACTOR Future<Void> sameVersionDifKey(Version ver, Key key, Optional<Value> value, Database cx,
+                              TransactionInfo info, TagSet tags) {
+	state ReadYourWritesTransaction tr(cx);
+
+	try {
+
+		tr.setOption( FDBTransactionOptions::READ_SYSTEM_KEYS );
+		state Optional<Value> valSS = wait(tr.get(key));
+		state Reference<WatchMetadata> metadata = cx->getWatchMetadata(key);
+
+		if (metadata.isValid() && valSS != metadata->value) { // val_3 != val_1
+			cx->deleteWatchMetadata(key);
+			
+			metadata->watchPromise.send(ver);
+			metadata->watchFutureSS.cancel();
+		}
+
+		if (valSS == value) { // val_3 == val_2
+			metadata = Reference<WatchMetadata>(new WatchMetadata(value, ver, info, tags));
+			cx->setWatchMetadata(key, metadata);
+
+			metadata->watchFutureSS = watchStorageServerResp(key, cx);
+		}
+
+		if (valSS != value) return Void(); // if val_3 != val_2
+
+		metadata = cx->getWatchMetadata(key);
+		if (metadata.isValid()) { // if val_3 == val_2
+			wait(success(metadata->watchPromise.getFuture()));
+		}
+
+		return Void();
+	}
+	catch(Error& e) {
+		state Error err = e;
+		wait(tr.onError(e));
+		throw err;
+	}
+}
+
+Future<Void> getWatchFuture(Version ver, Key key, Optional<Value> value, Database cx,
+                              TransactionInfo info, TagSet tags) {
+	Reference<WatchMetadata> metadata = cx->getWatchMetadata(key);
+
+	if (!metadata.isValid()) { // case 1: key not in map		
+		metadata = Reference<WatchMetadata>(new WatchMetadata(value, ver, info, tags));
+		cx->setWatchMetadata(key, metadata);
+
+		metadata->watchFutureSS = watchStorageServerResp(key, cx);
+		return success(metadata->watchPromise.getFuture());
+	} 
+	else if (metadata->value == value) { // case 2: val_1 == val_2
+		if (ver > metadata->version) {
+			metadata->version = ver;
+			metadata->info = info;
+			metadata->tags = tags;
+		}
+		
+		return success(metadata->watchPromise.getFuture());
+	} else if(ver > metadata->version) { // case 3: val_1 != val_2 && version_2 > version_1
+
+		cx->deleteWatchMetadata(key);
+
+		metadata->watchPromise.send(ver);
+		metadata->watchFutureSS.cancel();
+
+		metadata = Reference<WatchMetadata>(new WatchMetadata(value, ver, info, tags));
+		cx->setWatchMetadata(key, metadata);
+
+		metadata->watchFutureSS = watchStorageServerResp(key, cx);
+
+		return success(metadata->watchPromise.getFuture());
+	} else if (metadata->version == ver) { // case 5: val_1 != val_2 && version_1 == version_2
+		return sameVersionDifKey(ver, key, value, cx, info, tags);
+	}
+	// case 4: val_1 != val_2 && version_2 < version_1
+	return Void();
+}
+
+ACTOR Future<Void> watchValueMap(Future<Version> version, Key key, Optional<Value> value, Database cx,
+                              TransactionInfo info, TagSet tags) {
+	state Version ver = wait(version);
+	try {
+		Future<Void> watchFuture = getWatchFuture(ver, key, value, cx, info, tags);
+		wait(watchFuture);
+		return Void();
```

Comment:
You can collapse this and just return the result of `getWatchFuture`.

Reply:
Compiler doesn't allow this so will keep as is for now